lib/converters/scene-to-gltf.ts

```diff
@@ -1,17 +1,317 @@
+import { Buffer } from "node:buffer"
+import { mat4 } from "gl-matrix"
+import type { mat4 as Mat4ArrayType } from "gl-matrix"
+import type {
+  mat4 as GLMat4,
+  NodeIO as NodeIOClass,
+  Document,
+  Scene,
+  Transform,
+} from "@gltf-transform/core"
+
+import type {
+  Scene3D,
+  GLTFExportOptions,
+  ExternalModelInstance,
+  LoadedGLTFAsset,
+  CoordinateTransformConfig,
+} from "../types"
 import { GLTFBuilder } from "../gltf/gltf-builder"
+import { applyCoordinateTransform } from "../utils/coordinate-transform"
+
+interface MeshoptDecoderType {
+  ready: Promise<void>
+}
+
+interface DracoDecoderFactory {
+  createDecoderModule?: () => Promise<unknown>
+}
+
+type NodeIOConstructor = new () => NodeIOClass
+type MergeDocumentsFn = (target: Document, source: Document) => Map<unknown, unknown>
+type UnpartitionFn = () => Transform
+
+interface GltfToolkit {
+  NodeIO: NodeIOConstructor
+  mergeDocuments: MergeDocumentsFn
+  unpartition: UnpartitionFn
+  EXTMeshoptCompression: unknown
+  KHRDracoMeshCompression: unknown
+  meshoptDecoder?: MeshoptDecoderType
+  dracoDecoder?: unknown
+}
+
+let toolkitPromise: Promise<GltfToolkit> | null = null
 
 export async function convertSceneToGLTF(
   scene: Scene3D,
   options: GLTFExportOptions = {},
 ): Promise<ArrayBuffer | object> {
   const builder = new GLTFBuilder()
-
-  // Build the GLTF from our Scene3D structure
   await builder.buildFromScene3D(scene)
 
-  // Export to GLTF
-  const result = builder.export(options.binary)
+  if (!scene.externalModels || scene.externalModels.length === 0) {
+    return builder.export(options.binary)
+  }
+
+  const toolkit = await loadGltfToolkit()
+  const io = createNodeIO(toolkit)
+
+  const glbBuffer = builder.export(true) as ArrayBuffer
+  const baseDocument = await io.readBinary(new Uint8Array(glbBuffer))
+  const root = baseDocument.getRoot()
+
+  let targetScene = root.getDefaultScene()
+  if (!targetScene) {
+    const existingScenes = root.listScenes()
+    targetScene = existingScenes[0] ?? baseDocument.createScene("Scene")
+    root.setDefaultScene(targetScene)
+  }
+
+  for (const model of scene.externalModels) {
+    try {
+      await mergeExternalModel(baseDocument, targetScene, io, toolkit, model)
+    } catch (error) {
+      console.warn(`Failed to merge external model (${model.url}):`, error)
+    }
+  }
+
+  await baseDocument.transform(toolkit.unpartition())
+
+  if (options.binary) {
+    const binary = await io.writeBinary(baseDocument)
+    return uint8ArrayToArrayBuffer(binary)
+  }
+
+  const { json, resources } = await io.writeJSON(baseDocument)
+  embedResources(json, resources)
+  return json
+}
+
+async function loadGltfToolkit(): Promise<GltfToolkit> {
+  if (!toolkitPromise) {
+    toolkitPromise = (async () => {
+      const [core, functions, extensions] = await Promise.all([
+        import("@gltf-transform/core"),
+        import("@gltf-transform/functions"),
+        import("@gltf-transform/extensions"),
```

Comment:
If you keep PRing dynamic importants we’re going to assume you arent reading the code